core/src/main/java/overflowdb/storage/NodesWriter.java

```diff
@@ -0,0 +1,78 @@
+package overflowdb.storage;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import overflowdb.Node;
+import overflowdb.NodeDb;
+import overflowdb.NodeRef;
+
+import java.io.IOException;
+import java.util.Spliterator;
+import java.util.stream.StreamSupport;
+
+/**
+ * Persists collections of nodes in bulk to disk. Used either by ReferenceManager (if overflow to disk is enabled),
+ * or alternatively when closing the graph (if storage to disk is enabled).
+ */
+public class NodesWriter {
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final boolean isLogInfoEnabled = logger.isInfoEnabled();
+  private static final int LogAfterXSerializedNodes = 100_000;
+  private final NodeSerializer nodeSerializer;
+  private final OdbStorage storage;
+
+  public NodesWriter(NodeSerializer nodeSerializer, OdbStorage storage) {
+    this.nodeSerializer = nodeSerializer;
+    this.storage = storage;
+  }
+
+  /**
+   * writes all references to storage, blocks until complete.
+   */
+  public void writeAndClearBatched(Spliterator<? extends Node> nodes, int estimatedTotalCount) {
+    CountEstimates countEstimates = new CountEstimates();
+
+    StreamSupport.stream(nodes, false).forEach(node -> {
+      NodeDb nodeDb = null;
+      NodeRef ref = null;
+      if (node instanceof NodeDb) {
+        nodeDb = (NodeDb) node;
+        ref = nodeDb.ref;
+      } else if (node instanceof NodeRef) {
+        ref = (NodeRef) node;
+        if (ref.isSet()) nodeDb = ref.get();
+      }
+
+      if (nodeDb != null && nodeDb.isDirty()) {
+        try {
+          byte[] bytes = nodeSerializer.serialize(nodeDb);
+          storage.persist(ref.id(), bytes);
+          ref.clear();
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      if (isLogInfoEnabled) {
+        countEstimates.countEstimate++;
+        countEstimates.nodesToSerializeUntilLogging--;
+        if (countEstimates.nodesToSerializeUntilLogging == 0) {
+          float progressPercent = 100f * countEstimates.countEstimate / estimatedTotalCount;
+          countEstimates.nodesToSerializeUntilLogging = LogAfterXSerializedNodes;
+          logger.info(String.format("progress of writing nodes to storage: %.2f%s", Float.min(100f, progressPercent), "%"));
+        }
+      }
+    });
+  }
+
+  /**
+   * Counts aren't atomic or synchronized, but worst case we'll miss or duplicate a log.info message,
```

Comment:
This comment is misleading -- the counts don't need to be atomic or synchronized because they are not accessed concurrently.

If they were writtenn concurrently, in a relatively tight loop as above, then I'd advise to change something: We might know that the race is benign, but the CPU doesn't. More than one core accessing the same cacheline with at least one writer is always slow (correctness depends on atomic/volatile/etc, whether we access the same / overlapping addresses, and application logic; perf is always bad).

Reply:
I was using a parallel stream when I wrote that, and then had to change it to sequential due to race conditions. As noted above we'll go back to semi-parallel, which makes the comment accurate again I suppose. 

I don't mind so much about correctness here, but performance is important. 
My first naive thought was to batch the parallel streams, i.e. partition the nodes into chunks of e.g. 1024 and then use the same 'shared roughly correct racy counter'. That way we can scale down the "performance is bad" aspect of this. 
Do you have any suggestions?

Reply:
If we partition the node into chunks, then we can get the best of all worlds: Use an atomic counter (so no races), but only increment once a chunk has finished.

Alternatively we can use an atomic and assume that serialization+mvstore is so expensive that the (contended) atomic doesn't matter: Now that I think about it, this probably is the case.